Add "How to write a how-to" guide for developers

DragaDoncila opened the pull request:
This PR adds a template "How-To" document for developers to use when contributing their own documentation.

The guidelines in the template are based mainly on the [Diátaxis Framework](https://diataxis.fr/) and this helpful [Divio document](https://documentation.divio.com/how-to-guides/). 

The "prescribed" sections in this template are the introductory sentence, Prerequisites and Further Reading. I have tried to strike a balance between providing a standard template that helps users find consistent documentation that is easy to reference, while keeping the bulk of the document flexible enough to support a broad range of different types of "how-to". 

@melissawm I wasn't entirely sure where this document would fit into the bigger reorganization, so I've placed it here for now, but feel free of course suggest a better location for it.

Once we finalize this template, we will go through and update existing How-Tos to match this template. To get a concrete idea of what that would look like, ~I will open up a separate PR shortly migrating our existing installation guide to fit into this template~ see #184.

This is of course a draft, and all feedback is welcome. In particular, is a markdown document the best format for this or should it be a jupyter notebook?

DragaDoncila committed:
Add draft how-to guide

DragaDoncila commented on developers/contributing_how_tos.md:
```diff
@@ -0,0 +1,41 @@
+
+## Where does this how-to go?
+
+Once you've written your how-to, you should place this document under `getting_started/how_tos/` and open a PR in the [napari.github.io](https://github.com/napari/napari.github.io) repository.
```
We will probably need to update this to expand a bit more on where in the repository this document could go as I believe there are two sections for how-tos @melissawm?

melissawm replied:
Actually, I didn't anticipate a separate how-to section under the "Getting started" section. But I'm happy to discuss this if you feel it's needed

DragaDoncila replied:
No, I think I just got confused because the getting started section contains some how-tos. But I can see that this section is unlikely to be contributed to, so I guess we direct them to place their new document in a new folder we don't have yet called 'how-to's?

DragaDoncila commented on developers/contributing_how_tos.md:
```diff
@@ -0,0 +1,41 @@
+If you felt robbed of your chance to provide explanations, this is your opportunity to link out to useful references or discussions. You should also link to any relevant napari.org how-tos, tutorials or examples that can help the user learn more about the topic.
+
+- The [Diátaxis Framework](https://diataxis.fr/) is the basis for how we are organising our documentation
+- This [Divio document](https://documentation.divio.com/how-to-guides/) explains how-to guides in more detail
```
Once this is closer to being finalised we should also link to a good "how-to" example of our own as an exemplar.

DragaDoncila commented on developers/contributing_how_tos.md:
```diff
@@ -0,0 +1,41 @@
+## How-to vs. tutorial
+
+A how-to is a step by step guide designed to answer a specific "How do I..." question. This means the user already knows **what** they want to do, just not how. A how-to does not explain concepts or engage in discussion, and should prioritise practical usability. In comparison, a tutorial is a broader exploration, or guided tour of what is available. It might include detailed explanations and dis ...
+
```
We may also want some style guide suggestions here? I'm not sure if we want to be that prescriptive but for example the [Google documentation style guide](https://developers.google.com/style/) which they use over at [numpy](https://numpy.org/numpy-tutorials/content/tutorial-style-guide.html#make-use-of-the-google-doc-style-guide) would be an option

alisterburt replied:
good idea - this is good to highlight!

melissawm replied:
+1 for the google documentation style guide - I think it is very direct and informative without being over restrictive. They also have amazing [technical writing guides](https://developers.google.com/tech-writing) which I highly recommend.

GenevieveBuckley commented:
> This is of course a draft, and all feedback is welcome. In particular, is a markdown document the best format for this or should it be a jupyter notebook?

1. I would absolutely prefer a template in jupyter notebook format. 
    * So far, every single person I have talked to (with the possible exception of Talley) uses a notebook initially to make tutorials/examples/content. So that's the starting point people already have.
    * I think it's sufficient that if you look in the top right corner of the webpage you can click a button to download it as a jupyter notebook. I want users who stumble across the template in the github repository to *also* see it as a notebook. I want to be able to send someone a direct link and say "download this", instead of giving them a link and then a bunch of instructions for downloading in the correct format or converting to the notebook format.
2. The template should include a short bit about how to use `nbscreenshot` (nothing fancy, just that it exists and it's how we make screenshots). 
3. Optionally, I guess it could include a tiny bit more info on how you can select a specific slice or view in napari and how to hide this input cell, so the napari screenshot looks good (that's two things to learn, the napari programmatic controls, and how to tag the jupyter cells)
4. The template should include basic instructions on jupytext - what command do you run to turn your finished notebook into the text file that actually gets uploaded as a PR? Where do you open that PR, etc.

GenevieveBuckley commented:
I really want us to guide people through the whole process, rather than have unspoken knowledge about what happens between the gaps between the stages. I'd say the whole process includes:
1. Download this template notebook
2. Add your napari example, illustrating with screenshots
3. Check the final notebook runs end to end with no errors
4. Convert the notebook with jupytext
5. Open your PR in this reporitory (possibly linking to some git basics, for people who know image analysis but no git?)

GenevieveBuckley commented on developers/contributing_how_tos.md:
```diff
@@ -0,0 +1,41 @@
+
+## How-to vs. tutorial
+
+A how-to is a step by step guide designed to answer a specific "How do I..." question. This means the user already knows **what** they want to do, just not how. A how-to does not explain concepts or engage in discussion, and should prioritise practical usability. In comparison, a tutorial is a broader exploration, or guided tour of what is available. It might include detailed explanations and dis ...
```
I think it would be clearer to focus on what to do, or what makes a good how-to document.

It becomes more confusing for readers if there's a lot of:
* Sentences saying "don't do X or Y" (don't think of unicorn! ...now that's all you can think about, right?)
* Sentences trying to clarify the difference between a how-to and a tutorial

... you know, I think what I'm saying here is that you are inserting too much explanation content into your own how-to guide

DragaDoncila replied:
> ... you know, I think what I'm saying here is that you are inserting too much explanation content into your own how-to guide

Yes, I too had this feeling. I will reword to focus more on the how-to rather than how-not-to, and we can maybe work on disambiguation of tutorials and how-to's elsewhere

alisterburt replied:
+1, great suggestion @GenevieveBuckley - I think this section reads very nicely already if you just remove the 'don't do this' parts:

edit: maybe this should be moved to the top?
```suggestion

## How-to vs. tutorial

A how-to is a step by step guide designed to answer a specific "How do I..." question. This means the user already knows **what** they want to do, just not how.
```

melissawm replied:
I think this doesn't need to be removed, actually. I quite like the idea of making sure people understand this from the beginning. I'd suggest:

```suggestion
A how-to is a step by step guide designed to answer a specific "How do I..." question. This means the user already knows **what** they want to do, just not how. A how-to does not explain concepts or engage in discussion, and should prioritise practical usability.
```

This sentence:
```
In comparison, a tutorial is a broader exploration, or guided tour of what is available. 
It might include detailed explanations and discussion, and is usually aimed at users
who don't yet know what can be done.
```
could go into the tutorial exemplar/guide.

DragaDoncila commented:
> 2. The template should include a short bit about how to use `nbscreenshot` (nothing fancy, just that it exists and it's how we make screenshots).
> 3. Optionally, I guess it could include a tiny bit more info on how you can select a specific slice or view in napari and how to hide this input cell, so the napari screenshot looks good (that's two things to learn, the napari programmatic controls, and how to tag the jupyter cells)
> 4. The template should include basic instructions on jupytext - what command do you run to turn your finished notebook into the text file that actually gets uploaded as a PR? Where do you open that PR, etc.

@GenevieveBuckley yeah I wasn't sure about adding that much detail purely because not all `How-To`s will require running Python code (e.g. installation tutorial) and I wasn't sure if that was going to create a lot of overhead for developers who want to contribute a how-to that's mostly text and maybe an image or two.

>I really want us to guide people through the whole process, rather than have unspoken knowledge about what happens between the gaps between the stages.

Fully agree with this though. I just don't want to discourage anyone who wants to contribute something simpler. I guess we could do the detailed notebook-jupytext workflow in a separate document? But then is that too many documents?

@melissawm @neuromusic @chili-chiu what are your thoughts on balancing the level of detail for a simple how-to vs. a code heavy one?

GenevieveBuckley commented:
> @GenevieveBuckley yeah I wasn't sure about adding that much detail purely because not all How-Tos will require running Python code (e.g. installation tutorial) and I wasn't sure if that was going to create a lot of overhead for developers who want to contribute a how-to that's mostly text and maybe an image or two.


I think we can assume people contributing new tutoriials:
* Already know how to use python
* Already know how to use jupyter
* Already know how to use napari
* Already know some image analysis

And that they:
* Don't necessarily know how to use nbscreenshot
* Don't necessarily know how to use tags on jupyter cells to show/hide input
* Don't necessarily know how to use jupytext to convert their notebooks
* Don't necessarily know where to make a pull request (i.e https://github.com/napari/napari.github.io/pulls)
* Don't necessarily know how to make a pull request, or how to use git in general. I'd say most people contributing tutorials will know git, although there's currently some selection bias there now! We'd want to keep this very brief, maybe just include a link or two to more detailed guides elsewhere.

alisterburt commented on developers/contributing_how_tos.md:
```diff
@@ -0,0 +1,41 @@
+# How to contribute a napari How-To
+
+This should be a sentence (or brief paragraph) about what this how-to will achieve and how it helps address a common problem/task/need.
```
what the how-to will achieve doesn't feel super clear to me, is it about what the document itself will demonstrate? Maybe demonstrate is a clearer word (or I'm just slow 🙂)

melissawm replied:
This can be stated as what the reader will be able to do after reading through the document. I like to focus these statements in terms of the reader and not the document. So, I wouldn't say "what the how-to will achieve" but "what the reader will achieve after reading this how-to" (maybe too subtle but I feel it helps to pinpoint the goal of each document).

alisterburt commented on developers/contributing_how_tos.md:
```diff
@@ -0,0 +1,41 @@
+- this template
+- a list of what the user needs to prepare to be ready to start this how-to guide 
+- links to any existing napari.org tutorials or how-to guides that can help the user fulfill these prerequisites
+- the level of python/napari knowledge required to follow this how-to
+    - try to be specific about what skills are needed e.g. 
+    'connecting callbacks to layer events' or 'using matplotlib to produce plots'
```
super important, love this!

alisterburt commented on developers/contributing_how_tos.md:
```diff
@@ -0,0 +1,41 @@
+
+## How-to steps
+
+Headings in this section are yours to decide on. You should split your how-to guide into easy to follow, sequential steps to achieve the stated goal. Remember that your how-to may be used as a reference, so your headings should be descriptive and allow users to get a quick overview of the document, and which steps are relevant to them.
```
I like the idea of having an `outline` section before the actual steps in the how-to document. This section would quickly lists the steps, providing a super short summary of the steps to be taken and, if appropriate, a little more detail than the title of the section.

melissawm replied:
I think this will be automatically generated by the headings in the right-side toolbar, so not sure we need to include this again in the document itself.

alisterburt replied:
@melissawm you're right - I was initially thinking that for a longer form how-to this could be useful for providing a bit more detail but actually how-tos should probably remain very limited in scope

DragaDoncila replied:
@alisterburt I think that sort of outline/longer form intro would be perfect for a tutorial, but would expect how-tos to have limited enough scope for a brief paragraph to be sufficient (if a sentence isn't quite).

alisterburt commented on developers/contributing_how_tos.md:
```diff
@@ -0,0 +1,41 @@
+
+## Include pictures
+
+You should include pictures/videos, particularly when describing interactions with the viewer. If an action can be performed both through the viewer and through code, you should include both options.
```
<3

melissawm replied:
(should we include a mention of accessibility - "try to use descriptive and complete alt-text when including images"?)

melissawm commented:
> 4. The template should include basic instructions on jupytext - what command do you run to turn your finished notebook into the text file that actually gets uploaded as a PR? Where do you open that PR, etc.

Not to plug NumPy too much :) but we do have a short guide here:https://numpy.org/numpy-tutorials/content/pairing.html

melissawm commented:
> I think we can assume people contributing new tutorials:

These could be listed as pre-requisites, to make this assumption explicit to users reaching this page.

melissawm commented on developers/contributing_how_tos.md:
```diff
@@ -0,0 +1,41 @@
+
+## Further Reading
+
+If you felt robbed of your chance to provide explanations, this is your opportunity to link out to useful references or discussions. You should also link to any relevant napari.org how-tos, tutorials or examples that can help the user learn more about the topic.
```
+1, this is the way to go - delegate longer, deeper explanations elsewhere!

melissawm commented:
> @melissawm @neuromusic @chili-chiu what are your thoughts on balancing the level of detail for a simple how-to vs. a code heavy one?

We opted for heaving a separate tutorial about pairing the notebook. This could be a good option for you as well, since this will be applicable for how-tos, tutorials and maybe even explanation documents.

GenevieveBuckley commented:
> @GenevieveBuckley yeah I wasn't sure about adding that much detail purely because not all How-Tos will require running Python code (e.g. installation tutorial) and I wasn't sure if that was going to create a lot of overhead for developers who want to contribute a how-to that's mostly text and maybe an image or two.

Personally I really want to discourage content that is not runnable (it's way easier for text only content to drift out of date without anyone noticing). It also helps avoid the problem of instructions that are too vague, so people are less likely to get stuck at a step that's been glossed over.

DragaDoncila committed:
Updates from first round of reviews

DragaDoncila committed:
Remove horizontal rule

chili-chiu commented:
Perhaps at the end of the 'how-to' guide we can add a link to a simple real napari how-to example that's well-written? So that it's clear how it may look like by following the guide.

DragaDoncila committed:
Make guide executable

DragaDoncila commented:
I believe the latest changes have now addressed most of the feedback save for

1. linking to a good example (will be updating installation guide this weekend at the latest)
2. instructions for how to open a PR. I'm not sure how to proceed with that. I've looked for guides online that are sufficiently detailed, but they all assume some knowledge or other. How detailed do we want PR instructions to be? Do we want to write our own git/PR guide for napari? I don't mind doing it, I'm just not sure if we're not going a bit guide crazy - is teaching git something we want to take on as part of our core documentation?

melissawm commented:
My 2 cents: it is sufficient to link to the [github PR docs](https://docs.github.com/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request), which are also available in several different languages.

DragaDoncila commented:
> My 2 cents: it is sufficient to link to the github PR docs, which are also available in several different languages.

I was thinking roughly the same @melissawm with a note that if you have any trouble, reach out to us and we're more than happy to help with the PR process. @GenevieveBuckley does that seem reasonable to you, or would you prefer more detail around opening PRs and git?

GenevieveBuckley commented:
> > My 2 cents: it is sufficient to link to the github PR docs, which are also available in several different languages.
> 
> I was thinking roughly the same @melissawm with a note that if you have any trouble, reach out to us and we're more than happy to help with the PR process. @GenevieveBuckley does that seem reasonable to you, or would you prefer more detail around opening PRs and git?

Yes, I agree

GenevieveBuckley commented on developers/contributing_how_tos.md:
```diff
@@ -0,0 +1,63 @@
+
+# How to contribute a napari How-To
+
+This should be a sentence (or brief paragraph) about what the reader will achieve after following this
```
I think this phrasing is a bit too meta, like it's conflating this guide with a template in itself.
It reads more like "we meant to put content here at this placeholder but forgot", rather than the message you want to convey.

GenevieveBuckley commented:
Hi Draga! At the last docs meeting, you said you were blocked on reviews so I've just taken a look at the 3 PRs you mentioned.

As a general rule, I'd like us to minimize the number of dcouments as much as possible.
In my ideal world, there'd be one template you could download that also contained all of the instructions from start to finishing a new docs PR.
I understand that may not always be possible, but the more links you have to click to see all the information, the less chance people will see it all.

1. https://github.com/napari/napari.github.io/pull/184 looks good to me. I'd say we can merge this once you've finished tidying up any loose ends. 

2. I'd like https://github.com/napari/napari.github.io/pull/206 to also explain the very start of the process, how to download and start the template running with jupyter. Other than that, I don't have anything major to add.

3. https://github.com/napari/napari.github.io/pull/181 this one is probably the furthest away from being merge-able. 

In my opinion, the biggest blocker is this document needs to be a list of steps, which if you follow them means you end up with A new how to guide submitted to the napari documentation. Currently it feels half like we're trying to make this document a template also, and half like we're trying to work in all tips we have for best practices.

I don't disagree with any of the content that you have in there, it's all useful and good to know. But we're missing the main structure that needs to be here, which is the list of steps to follow in order to contribute something new to the docs.

Here are the two most relevant comment I've made:
- https://github.com/napari/napari.github.io/pull/181#issuecomment-930750803
- https://github.com/napari/napari.github.io/pull/181#issuecomment-930799697

DragaDoncila commented:
@GenevieveBuckley thanks for the detailed reviews. We discussed some of this at the meeting and came up with a structure that hopefully will minimize confusion for contributors and make this content easy to reference ([notes here](https://github.com/napari/meeting-notes/pull/16/files)). I will update all of the docs with the rest of your comments and will also open a draft PR with all of them together so we can preview the "workflow" as a whole.

GenevieveBuckley commented:
> I will update all of the docs with the rest of your comments and will also open a draft PR with all of them together so we can preview the "workflow" as a whole.

Ok, so is this now superseded by https://github.com/napari/napari.github.io/pull/283

DragaDoncila commented:
@GenevieveBuckley this is indeed superseded by #283, which will be closed and reopened in `napari/napari` once we are happy with the approach (it's in `napari.github.io` right now so we can preview easily.

I will close this PR now.
